desmos-labs/desmos
# Upgrade module implementation

## Description
This PR adds back the upgrade modules (closes #174).

You can test the integration running the following scripts.

<details>
<summary>1. Install cosmosd</summary>

```shell
mkdir -p ~/cosmosd && cd ~/cosmosd
git clone git@github.com:regen-network/cosmosd.git .
make build -o $GOBIN/cosmosd
```

</details>

<details>
<summary>2. Start the chain</summary>
    
```shell
echo "===> Deleting previous data"
rm -r -f $HOME/.desmosd
rm -r -f $HOME/.desmoscli
desmosd unsafe-reset-all

echo "===> Initializing chain"
desmosd init testchain --chain-id testchain

echo "===> Editing genesis.json"
sed -i 's/desmos/udaric/g' $HOME/.desmosd/config/genesis.json
sed -i 's/"voting_period": "172800000000000"/"voting_period": "120000000000"/g' \
  ~/.desmosd/config/genesis.json

echo "===> Creating genesis accounts"
desmosd add-genesis-account jack 100000000000000000000udaric
desmosd add-genesis-account desmos16f9wz7yg44pjfhxyn22kycs0qjy778ng877usl 10000000udaric

echo "===> Setting up CLI"
desmoscli config chain-id testchain

echo "===> Collecting genesis trasanctions"
desmosd gentx --amount 1000000000udaric --name jack
desmosd collect-gentxs

# For handling the upgrades
# go build -o $GOBIN/cosmosd

echo "===> Setting up upgrade manager"
mkdir -p ~/.desmosd/upgrade_manager/genesis/bin
mkdir -p ~/.desmosd/upgrade_manager/upgrades
cp $GOBIN/desmosd ~/.desmosd/upgrade_manager/genesis/bin


export DAEMON_NAME=desmosd
export DAEMON_HOME=$HOME/.desmosd
export DAEMON_RESTART_AFTER_UPGRADE=on

echo "===> Starting chain"
cosmosd start --pruning nothing
```  
</details>

<details>
<summary>3. Update the chain code</summary>

1. Update the chain code, adding the following lines to the `app.go` file:  
   ```go
   app.upgradeKeeper.SetUpgradeHandler("test", func(ctx sdk.Context, plan upgrade.Plan) {
   
   })
   ```

2. Install the update
   ```shell
   make build
   mkdir -p ~/.desmosd/upgrade_manager/upgrades/test/bin && cp build/desmosd ~/.desmosd/upgrade_manager/upgrades/test/bin
   ```

</details>

<details>
<summary>4. Submit, deposit and vote the upgrade proposal</summary>

```shell
echo "===> Submitting governance proposal"
desmoscli tx gov submit-proposal software-upgrade "test" \
  --title "test software upgrade proposal" \
  --description "something about the proposal here" \
  --deposit 100000000udaric \
  --upgrade-height=30 \
  --from jack \
  --yes --trace

echo "===> Waiting for transaction to be effective"
sleep 15

echo "===> Voting governance proposal"
desmoscli tx gov vote 1 yes --from jack --yes --trace
```
</details>

These will: 

1. Initialize a local chain
2. Create an upgrade proposal for an upgrade named `test` at block height `30`.
3. Deposit into the proposal and vote YES. 
4. Wait for the upgrade, and let `upgrade_manager` handle it

If everything works properly, this should be the output of the chain daemon at height 29-31: 

```
I[2020-07-21|09:14:23.942] Executed block                               module=state height=29 validTxs=0 invalidTxs=0
I[2020-07-21|09:14:23.954] Committed state                              module=state height=29 txs=0 appHash=8EC07260A92FD83F4D578C1A85644EFDBF425229F780AA4F858F3346AA817191
E[2020-07-21|09:14:29.045] UPGRADE "test" NEEDED at height: 30:         module=main 
I[2020-07-21|09:14:29.109] starting ABCI with Tendermint                module=main 
I[2020-07-21|09:14:29.456] applying upgrade "test" at height: 30        module=main 
I[2020-07-21|09:14:34.849] Executed block                               module=state height=31 validTxs=0 invalidTxs=0
I[2020-07-21|09:14:34.866] Committed state                              module=state height=31 txs=0 appHash=57DDD91E5F79D438B46A24088EF913B2BF62A49256154D5DC3AAAB152FF52D7F
I[2020-07-21|09:14:39.976] Executed block                               module=state height=32 validTxs=0 invalidTxs=0
```

If you would like to run everything from start again, execute everything from start.

## Checklist
- [x] Targeted PR against correct branch.
- [ ] Wrote integration tests (simulation & CLI). 
- [x] Added an entry to the `CHANGELOG.md` file.
